### ISSUE TITLE
Interesting interaction between -Wall/NamedFieldPuns and hlint

### ISSUE THREAD

**Issue by LeventErkok** [ndmitchell/hlint#1250]
```
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 9.0.1
$ hlint --version
HLint v3.3.1, (C) Neil Mitchell 2006-2021
```

This program compiles cleanly with `ghc -Wall -c a.hs`. (Note the `-Wall`):

```haskell
{-# LANGUAGE NamedFieldPuns #-}
module T where

newtype X = X { x :: Int }
       deriving Show

upd :: IO X
upd = do let x = 3
         pure X{x = x}
```

Unfortunately, `hlint` complains on it:

```
$ hlint a.hs
a.hs:1:1-31: Warning: Unused LANGUAGE pragma
Found:
  {-# LANGUAGE NamedFieldPuns #-}
Perhaps you should remove it.

1 hint
```

But if I remove the pragma, I get a warning from GHC for shadowing:

```
$ ghc -Wall -c a.hs
a.hs:8:14: warning: [-Wname-shadowing]
    This binding for ‘x’ shadows the existing binding
      defined at a.hs:4:17
  |
8 | upd = do let x = 3
  |              ^
```

Perhaps this is more of a GHC issue; I'm not sure why `NamedFieldPuns` avoids the shadow warning; but at the end of the day taking an `hlint` advice causes compilation with `-Wall` to be warning-free.

**Comment by shayne-fletcher** [comment]
g'day @LeventErkok. you've uncovered a ghc issue. i suggest raising an issue here https://gitlab.haskell.org/ghc/ghc/-/issues.

**Comment by LeventErkok** [comment]
@shayne-fletcher 

This was reported to GHC folks a while ago: https://gitlab.haskell.org/ghc/ghc/-/issues/18246

While I don't fully agree with GHC's behavior, the comments on that ticket suggest that this is the "correct" thing to do so far as GHC is concerned. So, the onus is on hlint in this case.

**Comment by shayne-fletcher** [comment]
i don't follow your reasoning as to why though. hlint is pointing out that `NamedFieldPuns` isn't being used. so you remove it as it's redundant and ghc now warns that there is shadowing. evidently you require that hlint ought not make that suggestion because it will lead to this warning. i don't believe i agree with this. i would instead say you now have a new lint to deal with, this time one thrown up by ghc rather than hlint.

**Comment by shayne-fletcher** [comment]
hmmm. on reading the ticket you referenced i guess there is an alternate point of view though. in that `NamedFieldPuns` IS being used (for its effect in allowing shadowing without warnings; similarly `RecordWildCards`). that is, it's not just punning that we might use `NamedFieldPuns` for.

**Comment by shayne-fletcher** [comment]
@LeventErkok 

after thinking about it for a bit, i'm not convinced by argument 2 and favor argument 1:

 - i say, `NamedFieldPuns` is not being used so it should be warned on by hlint
   - i think that shadowing without warning being valid in the presence of `NamedFieldPuns` is an implementation detail of that extension and not a motivation for enabling it; i say, if you don't pun you aren't using it and shouldn't enable it
 - removing `NamedFieldPuns` restores warning on shadowing which gives rise to another lint that should then be addressed and that's fine

regardless of my opinion on that, say @ndmitchell decides hlint should only warn on `NamedFieldPuns` if there's no punning and no shadowing? sadly i rather suspect that this is beyond hlint's capacity to implement (or at least, "easily" implement). detecting shadowing requires renaming and hlint operates strictly on parse trees at this time.

**Comment by LeventErkok** [comment]
@shayne-fletcher 

I agree this is rather unfortunate; and perhaps can be marked "won't-fix."

My mental model of hlint is that whatever it recommends I can apply it and GHC will still be happy. This is true to a large extent. It would be terrible if GHC is happy but the semantic has changed by the advice; but if the advice causes GHC to complain that's not all that terrible.

Feel free to close the ticket as you see fit; but perhaps it's worth a mention in the code somewhere.
